pkg/assembler/backends/inmem/pkg.go

```diff
@@ -421,20 +436,58 @@ func (c *demoClient) IngestPackage(ctx context.Context, input model.PkgInputSpec
 	return c.buildPackageResponse(collectedVersion.id, nil)
 }
 
-func duplicatePkgVer(versions pkgVersionList, input model.PkgInputSpec) (bool, *pkgVersionNode) {
-	for _, v := range versions {
-		if noMatchInput(input.Version, v.version) {
-			continue
-		}
-		if noMatchInput(input.Subpath, v.subpath) {
-			continue
-		}
-		if !reflect.DeepEqual(v.qualifiers, getQualifiersFromInput(input.Qualifiers)) {
-			continue
+// hash the canonical representation of a version
+func hashPkgVersionNode(version *pkgVersionNode) (string, error) {
+	if version == nil {
+		return "", fmt.Errorf("version is nil")
+	}
+	return hashVersionHelper(version.version, version.subpath, version.qualifiers)
+}
+
+// hash the canonical representation of a version
+func hashPkgInputSpecVersion(input model.PkgInputSpec) (string, error) {
+	qualifiers := getQualifiersFromInput(input.Qualifiers)
+	return hashVersionHelper(nilToEmpty(input.Version), nilToEmpty(input.Subpath), qualifiers)
+}
+
+func hashVersionHelper(version string, subpath string, qualifiers map[string]string) (string, error) {
+	hash := sha256.New()
+
+	_, err := hash.Write([]byte(version))
+	if err != nil {
+		return "", fmt.Errorf("Error writing when computing the hash of a version: %v", err)
+	}
+	_, err = hash.Write([]byte(subpath))
+	if err != nil {
+		return "", fmt.Errorf("Error writing when computing the hash of a version: %v", err)
+	}
+
+	qualifierSlice := []string{}
+	for key, value := range qualifiers {
+		qualifierSlice = append(qualifierSlice, fmt.Sprintf("%s:%s", key, value))
+	}
+	slices.Sort(qualifierSlice)
+	for _, qualifier := range qualifierSlice {
+		_, err = hash.Write([]byte(qualifier))
+		if err != nil {
+			return "", fmt.Errorf("Error writing when computing the hash of a version: %v", err)
 		}
-		return true, v
 	}
-	return false, nil
+
+	digest := hash.Sum([]byte{})
```

Comment:
it seems like the key that is being hashed may not be unique. For example, if we have:

version: "v1"
subpath: ".2-abc"

vs 

version: "v1.2-abc"
subpath: ""

Reply:
May consider using `fmt.Sprintf` to construct the string first and then hash it, will be a little less efficient but more readable